tensorflow/compiler/xla/service/all_reduce_reassociate.cc

```diff
@@ -37,9 +40,57 @@ bool AreCompatible(const HloAllReduceInstruction *ar0,
   std::optional<AllReduceKey> key0 = GetAllReduceKey(ar0);
   std::optional<AllReduceKey> key1 = GetAllReduceKey(ar1);
   auto kind0 = MatchReductionComputation(ar0->to_apply());
+
   return key0 && key1 && kind0 && *key0 == *key1 && kind0 == op_kind;
 }
 
+bool AreCompatibleConverts(const HloInstruction *convert0,
+                           const HloInstruction *convert1) {
+  bool is_compatible = true;
+  // For numerical stability, we only re-order ops with casts from a narrow type
+  // to a wider type.
+  if (convert0) {
+    is_compatible &=
```

Comment:
Would it be better to use `primitive_util::CastPreservesValues` here? That will also filter out casts like f16<->i32 which are not value preserving even if the bit width is f16 < bitwidth of i32

Reply:
makes sense, updated to use CastPreservesValues